### ISSUE TITLE
PRD: Build a recursive semantic graph-presentation judge

### ISSUE THREAD

**Issue by vishaltandale00** [vishaltandale00/relayer-graphcomplete#192]
## Summary

Replace the simulated-user judge's flat, independently declared recursive-structure ratings with an LLM-authored recursive judge tree.

The judge receives tools for reviewing the accepted graph bottom-up. Each expansion layer produces one fixed-size score vector and one aligned semantic vector. A parent node consumes the complete results of its action layers as signals, semantically compresses them into its own node result, and contributes that result to its parent layer. The completed root `LayerResult` becomes the recursive graph input to the final turn judgment.

This PRD changes only Eval judging and judge inspection. It does not change graph semantics, graph authoring, harness prompts, model execution, or accepted product output.

## Problem

The current graph-presentation judge asks the model to independently classify expansion, reference, and invoke need as `none`, `helpful`, or `required`. The review store constrains missing required disclosure, but it permits a node with every absent action classified as `none` to receive recursive-disclosure rating 4.

This creates two failures:

1. The judge can rationalize an action-free graph after seeing that no actions exist instead of comparing the authored allocation with plausible `expand`, `reference`, `invoke`, and `stop` choices.
2. Deep layer quality and semantic failures do not return through the expansion graph as structured signals that can change the scores and explanations of earlier layers.

The completed ofetch evaluations expose the gap:

- `run-2026-08-26T04-11-26-245Z-b0f3ab1d` received presentation 4 for one three-node root layer with no node actions, despite not explaining the material implementation work.
- `run-2026-08-26T13-39-18-463Z-521611a1` improved the visible handoff to four substantive nodes, but again received recursive-structure 4 on every node after the judge classified all absent actions as unnecessary.

The product needs an inspectable recursive judgment, not a fixed node-count requirement, a mandatory expansion rule, or a literal numeric formula.

## Product outcome

1. The judge builds one recursive review tree over expansion layers.
2. Every graded layer returns one bounded `LayerResult` containing fixed-size node scores, aligned node semantics, a layer summary, and evidence links.
3. Parent nodes receive complete child `LayerResult` objects as semantic signals and compress them into their own node result.
4. The judge explicitly ranks `expand`, `reference`, `invoke`, and `stop` before grading the corresponding actual action or destination.
5. Multiple actions are handled as multiple layer-result signals; the judge keeps evaluating allocation until `stop` is preferred.
6. Selection quality and destination delivery remain separate.
7. Ordinary deep weaknesses decay locally as results cross expansion layers, while semantic findings that undermine a parent promise are reinterpreted as parent-level findings.
8. The final turn judgment consumes the original request, artifact evidence, and root `LayerResult`, rather than separately aggregating every descendant again.
9. Relayer Eval persists and renders the full score and semantic tree for inspection.

## Recursive judgment model

The semantic function is:

```text
gradeLayer(layer):
    recursively grade every expansion layer reached from its nodes

    for each node:
        grade visible node content
        rank expand, reference, invoke, and stop
        compare the preferred choice with the authored allocation
        grade each authored action
        consume expansion and reference LayerResults as signals
        keep invoke delivery and recursion unavailable
        compress all signals into one node score and semantic summary

    compress the layer's node results into one LayerResult
    return that LayerResult to the layer above
```

Evaluation is bottom-up:

```text
deep expansion layer
    -> child LayerResult
    -> source node compression
    -> parent LayerResult
    -> parent source-node compression
    -> root LayerResult
    -> final turn judgment
```

There is no separate scheduler or graph kernel. This is a judge-side traversal and review artifact over immutable accepted graph state.

## LayerResult contract

Every expansion-reachable layer produces exactly one result:

```ts
interface LayerResult {
  readonly layerId: string;
  readonly depth: number;

  // ADR 0005 permits at most eight visible nodes per layer.
  // Empty slots are explicit nulls and node IDs preserve slot identity.
  readonly nodeScores: readonly [
    NodeScore | null,
    NodeScore | null,
    NodeScore | null,
    NodeScore | null,
    NodeScore | null,
    NodeScore | null,
    NodeScore | null,
    NodeScore | null,
  ];

  readonly nodeSemantics: readonly [
    NodeSemanticSummary | null,
    NodeSemanticSummary | null,
    NodeSemanticSummary | null,
    NodeSemanticSummary | null,
    NodeSemanticSummary | null,
    NodeSemanticSummary | null,
    NodeSemanticSummary | null,
    NodeSemanticSummary | null,
  ];

  readonly layerRatings: LayerRatings;
  readonly layerSummary: string;
  readonly evidence: readonly ScreenshotEvidenceRef[];
}
```

The exact tuple syntax may be represented as a runtime-validated length-eight array. The invariant is fixed capacity and aligned node-score/semantic slots, not a TypeScript implementation preference.

### Node score channel

Each occupied node slot preserves four dimensions:

```ts
interface NodeScore {
  readonly nodeId: string;
  readonly content: 1 | 2 | 3 | 4;
  readonly actionAllocation: 1 | 2 | 3 | 4;
  readonly actionDelivery: 1 | 2 | 3 | 4 | null;
  readonly recursiveQuality: 1 | 2 | 3 | 4 | null;
}
```

- `content` grades the visible node in its layer context.
- `actionAllocation` grades whether the authored set of expand/reference/invoke actions and implicit stop was appropriate.
- `actionDelivery` grades assessable expansion and reference delivery. It is `null` when no destination delivery is assessable.
- `recursiveQuality` grades expansion-child graph quality. It is `null` for terminal, reference-only, and unresolved invoke paths.

The LLM receives child vectors and semantic summaries as signals. It performs semantic compression; the product must not replace this judgment with an arithmetic mean or a literal multiplicative formula.

### Semantic channel

Every occupied score slot has an aligned semantic summary that explains the numbers and preserves evidence:

```ts
interface NodeSemanticSummary {
  readonly nodeId: string;
  readonly meaning: string;
  readonly delivered: string;
  readonly limitations: string;
  readonly effectOnLayer: string;
  readonly evidence: readonly ScreenshotEvidenceRef[];
}
```

The LLM may compress several child messages into one parent explanation. No deterministic obligation ledger, deduplication algorithm, or keyword propagation system is introduced. The model receives the child results and decides what they mean for the current node and layer.

## Sequential action-allocation judgment

At each allocation step, the judge first evaluates the source node state and records a full ranking of:

```text
expand
reference
invoke
stop
```

It records the important semantic margin between the preferred choice and the authored choice:

- `close`: the authored choice is a reasonable alternative;
- `clearly_better`: the preferred choice would meaningfully improve the graph;
- `necessary`: the preferred choice is needed to avoid a material gap.

The ranking is qualitative semantic-logit evidence, not a probability distribution.

The judge then grades the actual action separately:

1. Was its action type an appropriate allocation from the source node?
2. Is its label and placement clear?
3. For expand/reference, does its layer result deliver what the action promises?
4. For expand, what does the recursively graded child graph contribute?

After consuming an actual action's delivered information, the judge evaluates the next allocation again. When `stop` becomes preferred, the ideal allocation ends and every unused authored action is an extra allocation.

The judge applies asymmetric judgment:

- a useful nonessential extra action may remain compatible with 4;
- a clearly unnecessary extra action normally creates a local allocation weakness;
- a missed action separated by `clearly_better` is a missed useful action;
- a missed action separated by `necessary` is a missed required action;
- under-allocation is more serious than comparable over-allocation.

The worst meaningful allocation error controls the node's `actionAllocation` dimension. Strong actions remain visible in the semantic channel and are not erased by the compressed score.

### Expand

- Allocates deeper explanation now.
- Its target layer is recursively graded.
- Depth decay primarily lowers the default preference for eager expansion; it does not lower the judged importance or latent information density of the source node.
- Expansion may remain preferred at any depth when its semantic value overcomes allocation and fragmentation cost.

### Reference

- Reuses material whose `LayerResult` was produced through its own expansion path.
- Reference traversal never triggers recursive grading.
- The source node consumes the existing target `LayerResult` as a signal and judges relevance, inspectability, and support.
- Reference preference is mostly depth-neutral because it reuses rather than authors explanatory structure.
- Reference cycles do not create judge recursion.

### Invoke

- Offers deferred user-controlled allocation as an alternative to eager expansion.
- It requires a concrete description of the new work it would produce; depth alone cannot justify it.
- The judge grades allocation, placement, label, clarity, and apparent value.
- The current review does not recurse into invoke because its result layer does not exist yet.
- Invoke delivery and recursive scores remain explicitly `null`; absence of a result cannot create earned credit.

### Stop

- Is the implicit end of a node's allocation sequence.
- It competes with expand, reference, and invoke at every step.
- It may be preferred for an atomic root node or a deep node whose remaining actions have insufficient marginal value.
- Once preferred, unused authored actions are graded as extra allocation.

## Recursive compression and depth decay

Depth decay applies locally at each expansion boundary:

```text
Layer N result
    -> locally decayed compression into Layer N-1
    -> locally decayed compression into Layer N-2
```

The model uses depth, importance, information density, existing graph allocation, duplication, and navigation cost as reasoning signals. No fixed depth threshold or equal-share formula is introduced.

If a deep finding remains local, its ordinary influence decays through recursive compression. If it undermines the parent action's promise, the parent LLM rewrites it as a parent-level semantic finding and reflects it in the parent node's own score and semantic slot. That new local judgment can propagate again through the same recursive function.

Multiple action-layer vectors are supplied to the LLM together. The LLM decides their collective effect using semantic importance and the worst meaningful deficit rather than an arithmetic mean.

## Judge tool and review-store behavior

The simulated-user judge tools and review store must support the recursive function without exposing graph mutation:

- inventory supplies exact expansion depth, source node/layer provenance, action relation, and destination IDs;
- screenshot and interaction tools remain the authority for visible graph communication;
- the judge can write or revise node scores, node semantics, action-allocation rankings, and layer results;
- a layer cannot be finalized until all of its expansion-child layers have final `LayerResult` records;
- reference targets reuse already finalized results and create no recursive coverage requirement;
- unresolved invoke actions require no destination review;
- every expansion-reachable layer is finalized exactly once;
- every visible node occupies exactly one aligned score/semantic slot;
- every authored node action is reviewed, including actions remaining after `stop` becomes preferred;
- the root turn review can finalize only after the root `LayerResult` exists;
- judge revisions remain inspectable and immutable final review artifacts retain their complete tree.

Tool names and request partitioning may follow the existing `reviewLayer`, `reviewNode`, and `submitReview` conventions or introduce a versioned replacement. The product contract is the persisted recursive result and enforced review order, not a specific tool name.

## Final turn judgment

The final turn judge consumes only:

```text
original request
candidate artifact evidence
root LayerResult
```

It does not separately aggregate every descendant a second time. Deep quality and semantic problems have already been compressed through the recursive graph.

The turn review retains the existing separation between task-outcome grading and graph-presentation grading. Verifiable work quality cannot be compensated for by a polished graph, and graph quality cannot be inferred from passing verifier receipts.

## Inspection experience

Relayer Eval must expose the complete finalized judge tree:

- root and expansion layers are navigable in recursive review order;
- every layer shows its eight aligned node score and semantic slots;
- every node shows the four score dimensions and semantic explanation;
- action allocation shows the four-choice ranking, actual choice, margin, and selection finding;
- expansion actions link to their recursively graded child `LayerResult`;
- reference actions link to the reused target result without duplicating it;
- invoke actions show action-only judgment and `null` delivery/recursive values;
- evidence links open the screenshots used by the judge;
- the root turn judgment shows the root result it consumed.

The stored JSON artifact remains authoritative. The UI is an inspector for that artifact, not a second scoring implementation.

## Acceptance criteria

- [ ] The judge contract persists one inspectable recursive tree per completed turn review.
- [ ] Every expansion-reachable layer produces exactly one finalized `LayerResult`.
- [ ] Every `LayerResult` contains aligned, runtime-validated eight-slot node score and semantic vectors.
- [ ] Every occupied node score contains content, allocation, delivery, and recursive-quality dimensions with correct nullability.
- [ ] The judge records a full `expand` / `reference` / `invoke` / `stop` ranking and `close` / `clearly_better` / `necessary` margin for each allocation step.
- [ ] The judge ranks from source-node state before grading actual destination delivery.
- [ ] Multiple actions and multiple same-kind actions are accepted as multiple semantic signals and every authored action is reviewed.
- [ ] `stop` ends ideal allocation and remaining authored actions are inspected as extras.
- [ ] Useful extras can remain compatible with 4; unnecessary extras are penalized less than missed required actions.
- [ ] Expansion children are recursively graded bottom-up.
- [ ] Reference targets are never recursively regraded and reuse their existing `LayerResult`.
- [ ] Invoke actions receive action-only review with `null` delivery and recursive values.
- [ ] Parent node results consume all available action-layer score and semantic vectors as LLM signals.
- [ ] Local depth decay occurs at each expansion compression step without a fixed depth cutoff or equal-share assumption.
- [ ] Deep semantic findings can be rewritten into parent-level findings when they undermine the parent promise.
- [ ] The final turn review consumes the original request, artifact evidence, and root `LayerResult` without separately reaggregating descendants.
- [ ] Relayer Eval renders the complete tree, vectors, semantic summaries, allocation rankings, evidence, and reused reference links.
- [ ] Review validation rejects incomplete expansion coverage, misaligned vectors, unknown IDs, invalid score/null shapes, missing authored-action reviews, and a final turn without a root result.
- [ ] Deterministic fixture tests cover flat answers, multi-action nodes, repeated action kinds, nested expansions, reference reuse and cycles, unresolved invokes, premature stop, useful extras, unnecessary extras, missed useful actions, missed required actions, failed child delivery, local deep weaknesses, and parent-impacting deep findings.
- [ ] Default tests use fixture judges and make no paid inference calls.
- [ ] `npm run check` passes.
- [ ] `npm run build` passes.

## Non-goals

- Changing any authoring-agent or harness prompt.
- Requiring an action on every node.
- Requiring a minimum number of nodes, layers, or expansion depth.
- Introducing a literal multiplicative score formula, numeric semantic logits, probabilities, or an online-learning bandit.
- Introducing a deterministic obligation ledger or semantic deduplication algorithm.
- Regrading a layer merely because it is reached by reference.
- Executing invoke actions during immutable review.
- Changing expansion, reference, invoke, acceptance, or graph authority semantics.
- Adding another recursive execution scheduler.
- Rewriting historical Eval bundles.
- Establishing final calibration thresholds or benchmark promotion gates in this slice.
- Modifying task-outcome verifier contracts.

## Evidence and compatibility

- Historical bundles remain immutable and readable under their recorded rubric versions.
- The new recursive judge tree uses a new contract and rubric version.
- Existing screenshot capture, artifact inspection, read-only sandboxing, and review evidence authority remain unchanged.
- Existing outcome-grade and presentation-grade separation remains unchanged.
- Flat answers remain valid when the judge prefers `stop`.
- The current ofetch runs are motivating evidence, not score-locking acceptance fixtures for this slice.
